packages/big-design/src/components/Tree/reducer.utils.ts

```diff
@@ -0,0 +1,181 @@
+  const initialState: TreeState<T> = {
+    nodeMap: new Map<TreeNodeId, MapValues>(),
+    expandedNodeIds: new Set(),
+    flattenedNodeIds: [],
+    focusedNode: null,
+    selectedValues: new Set(),
+    visibleNodeIds: [],
+  };
+
+  return recursiveInitialize(nodes, initialState, radio);
+};
+
+const recursiveInitialize = <T>(
+  nodes: TreeProps<T>['nodes'] | undefined,
+  state: TreeState<T>,
+  radio: boolean,
+): TreeState<T> => {
+  if (!nodes || nodes.length < 1) {
+    return state;
+  }
+
+  // Order of statements in this loop is important.
+  nodes.forEach((node) => {
+    state.flattenedNodeIds.push(node.id);
+
+    state.nodeMap.set(node.id, {
+      children: new Set(node.children?.map((child) => child.id)),
+      id: node.id,
+      parent: getParentId(state.nodeMap, node.id),
+    });
+
+    if (state.focusedNode === null) {
+      state.focusedNode = node.id;
+    }
+
+    if (node.expanded) {
+      state.expandedNodeIds.add(node.id);
+    }
+
+    const parent = state.nodeMap.get(node.id)?.parent;
+
+    if ((parent !== undefined && state.expandedNodeIds.has(parent)) || parent === undefined) {
+      state.visibleNodeIds.push(node.id);
+    }
+
+    if (node.children && node.children.length > 0) {
+      recursiveInitialize(node.children, state, radio);
+    }
+
+    // Needs to happen after the recursive call for radio Trees
+    // We want the closest node to the root to be selected.
+    if (node.selected && node.value) {
+      if (radio) {
+        state.selectedValues.clear();
+      }
+
+      state.focusedNode = node.id;
+      state.selectedValues.add(node.value);
+    }
+  });
+
+  return state;
+};
+
+const getParentId = (nodeMap: Map<TreeNodeId, MapValues>, id: TreeNodeId) => {
+  const iterator = nodeMap.entries();
+
+  for (const [, value] of iterator) {
+    if (value.children && value.children.has(id)) {
+      return value.id;
+    }
+  }
+
+  return undefined;
+};
+
+export const getNextVisibleNode = <T>(state: TreeState<T>, id: TreeNodeId): TreeNodeId => {
+  const index = state.visibleNodeIds.indexOf(id);
+
+  if (index !== -1 && index + 1 < state.visibleNodeIds.length) {
+    return state.visibleNodeIds[index + 1];
+  }
+
+  return id;
+};
+
+export const getPreviousVisibleNode = <T>(state: TreeState<T>, id: TreeNodeId): TreeNodeId => {
+  const index = state.visibleNodeIds.indexOf(id);
+
+  if (index !== -1 && index - 1 >= 0) {
+    return state.visibleNodeIds[index - 1];
+  }
+
+  return id;
+};
+
+export const toggleNode = <T>(state: TreeState<T>, action: Extract<Action<T>, { type: 'TOGGLE_NODE' }>) => {
+  const node = state.nodeMap.get(action.id);
+
+  if (state.expandedNodeIds.has(action.id)) {
+    state.expandedNodeIds.delete(action.id);
+
+    node?.children.forEach((childId) => {
+      const index = state.visibleNodeIds.indexOf(childId);
+
+      if (index > -1) {
+        state.visibleNodeIds.splice(index, 1);
```

Comment:
`splice` actually mutates the original array. You might want to treat `state` as immutable, because I think `state` refers to a state that's passed in from a `reducer`. I'd return a copy instead.

Reply:
My only quandary with that is since is that we are using `Set`'s. When we `.add` or `.delete` on a set it's mutating the `Set`, so why should we not do it consistently for our other state objects?

Reply:
Yes, what I meant is, you should treat all objects as immutable (especially when you're using the reducer pattern) - whenever they are passed in as an argument of a method or function. So with `Set`, you should create a new one as well. It is okay to mutate an object if it is created inside the same function / method scope. But once it is passed to another scope, it shouldn't be mutated.

Reply:
I started refactoring it, I'll hopefully have a commit soon for it.